Redirect to the previous screen when back button is clicked, even when images are uploading

anitaa1990 opened the pull request:
Fixes #4333 .

**This PR is part of a work in progress feature so it is merged against a feature branch. It is in draft till #4432 can be reviewed and merged**

### Changes
- Allows the merchant to redirect to the Product detail screen when the back button is clicked, when images are being uploaded.
- If an image upload fails in the Product detail or Variation detail screen, a similar Snackbar would be displayed with a `Details` button. 
- Clicking on the `Details` would redirect to a new screen that displays the list of images that failed to upload.

### Screenshots
##### Product Detail redirection
<img src="https://user-images.githubusercontent.com/22608780/125621739-6626e6ce-49ad-4d4c-a0bc-7812facada02.gif" width="300"/>

##### Variation Detail redirection
<img src="https://user-images.githubusercontent.com/22608780/125621939-1f1f245c-8436-4671-87a5-05a9a8549528.gif" width="300"/>

### Testing
- Try to upload an image with an extension, other than `png, jpeg, jpg or gif`. There is local validation in FluxC so any images without these extensions will result in an error.
- While the image is being uploaded, click on the back button. You should be redirected to the Product Details screen.
- Verify that the error snackbar is displayed with a DETAILS button. Click on the `Details` button and note that you are redirected to a new screen that displays a list of errors.
- Follow the above steps for a variation of a product.

Update release notes:

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

anitaa1990 committed:
Redirect to the previous screen when back button is clicked, even when images are uploading

anitaa1990 committed:
Added logic to display error snackbar in Product detail & variation detail screens

anitaa1990 committed:
Fixed failing unit tests

anitaa1990 committed:
Added redirection logic for displaying error list from variation detail screen

anitaa1990 committed:
Fixed failing unit test

anitaa1990 committed:
Fixed failing unit test (again)

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/base/UIMessageResolver.kt:
```diff
@@ -57,7 +57,7 @@ interface UIMessageResolver {
         actionListener: View.OnClickListener
     ): Snackbar {
-        return getSnackbarWithAction(
+        return getIndefiniteSnackbarWithAction(
```
I wonder if we shouldn't rename the parent method `getActionSnack` as well to make it clear it's a persistent snackbar?

anitaa1990 replied:
👍 Fixed in 467a5e0

hichamboushaba reviewed (commented):
Great job Anitaa, the experience is way better.

But I found a bug, I can't remove any image from the products:
<img width=360 src="https://user-images.githubusercontent.com/1657201/125934342-90630ca9-7797-4594-aa6d-bfb7ffd55a9b.gif"/>

I believe the cause is the change from `ExitWithResult` to `Exit` in the images fragment 🤔

anitaa1990 committed:
Added logic to refresh product detail images if image is added/deleted

anitaa1990 committed:
Added unit tests for the exitWithResult event

anitaa1990 committed:
Updated getActionSnack method name

anitaa1990 commented:
Nice catch @hichamboushaba! The product images in the detail screen were not getting refreshed, if images were added as well as deleted 🤦‍♀️ .

I fixed that in 3adee9c and also added unit tests for that. Thanks!

anitaa1990 commented:
I'm removing your assignment since your afk this week 👍

wzieba commented:
Thanks @anitaa1990 for those improvements, like @hichamboushaba mentioned experience is indeed better!

I've found one issue though, it's about replacing image on variations:


https://user-images.githubusercontent.com/5845095/126616397-1fbe1cf8-81e2-4fe5-8c81-3e5197941eef.mov

The gallery view in product details seems to be not updated after image upload fail.

It reminds me of the issue #3755 but I don't know exactly how they relate to each other. Do you think it's in scope of this PR?

anitaa1990 commented:
Sorry for the late reply @wzieba! Thank you for the review! 

> The gallery view in product details seems to be not updated after image upload fail.

I am able to replicate this issue in `develop` as well so think it is unrelated to the changes in this PR. Would you be able to confirm if you are able to reproduce this in the current release or `develop` as well? If so, I do think we can merge this PR as is and work on fixing that issue in #3755

wzieba commented:
Yes, `develop` has similar issue - it's not 100% same (as we can't navigate back without saving photo update) but very similar.

> I do think we can merge this PR as is and work on fixing that issue in #3755

Sure - I'll update #3755 with this additional info!

PR itself looks good to me! :shipit:
